### PR TITLE
handling of future and config default wrapper

### PR DESCRIPTION
**Pull request by BenGriffith**
Feel free to review if you would like to but tagging you as an FYI that merging changes discussed in previous comments.

- handling of future objects by writing to `pubsub.log`
  - updated so that logging to `pubsub.log` only occurs when `--pubsub` is enabled
- implemented the config default wrapper

### CODE REVIEW

File: tollway/utils.py [hunk 1]
```diff
@@ -1,43 +1,70 @@
...
 from pathlib import Path
-from typing import Mapping, TypedDict, Union
+from typing import Callable, Mapping, TypedDict, Union
 
 from decouple import config
 from google.cloud import pubsub_v1
+from google.cloud.pubsub_v1.publisher.futures import Future
 from google.oauth2 import service_account
 
+Topic = namedtuple("Topic", ["publisher", "topic_path"])
```
Comment (bbelderbos) [inline tollway/utils.py:12]: I prefer typed ones these days, see https://codechalleng.es/tips/define-a-namedtuple-with-type-hints

File: tests/conftest.py [hunk 2]
```diff
@@ -148,8 +142,8 @@ def original_and_late_events_timestamps(generate_event_pair):
 def duplicate_event(past_events, pubsub):
     event = DuplicateEventProcessor(
         events_log=past_events,
-        publisher=pubsub[0],
-        topic_path=pubsub[1],
+        publisher=pubsub.publisher,
+        topic_path=pubsub.topic_path,
```
Comment (bbelderbos) [inline tests/conftest.py:146]: nice

File: tollway/constants.py [hunk 3]
```diff
@@ -2,14 +2,14 @@
...
-from tollway.utils import get_envs
+from tollway.utils import get_config_with_default
 
 STATE_NAMES = [state.name for state in STATES_AND_TERRITORIES]
 URL_PATH = "wiki/List_of_toll_roads_in_the_United_States"
 TOLLWAYS_URL = f"https://en.wikipedia.org/{URL_PATH}"
 TIMESTAMP_FORMAT = "%Y-%m-%d %H:%M:%S.%f %z"
-DUPLICATE_RATE = get_envs("duplicate")
-ALL_EVENTS_COUNT = get_envs("all")
+DUPLICATE_RATE = get_config_with_default(key="DUPLICATE_RATE", default_value="50")
```
Comment (bbelderbos) [inline tollway/constants.py:11]: should these be ints?

File: tollway/utils.py [hunk 4]
```diff
@@ -1,43 +1,70 @@
...
-        return publisher, topic_path
-    return None, None
+        return Topic(publisher=publisher, topic_path=topic_path)
+    return Topic(publisher=None, topic_path=None)
+
+
+def get_pubsub_logger(pubsub: bool) -> logging.Logger:
+    logger = logging.getLogger("pubsub")
+    if pubsub:
+        logging.basicConfig(
```
Comment (bbelderbos) [inline tollway/utils.py:35]: maybe set this up at the module or even package (__init__.py) level?

File: tollway/utils.py [hunk 5]
```diff
@@ -1,43 +1,70 @@
...
-def get_envs(env: str) -> int:
-    defaults = {"duplicate": 50, "all": 250, "seconds": 10, "minutes": 20, "hours": 30, "days": 100}
-
-    env_variable = config(env.upper(), default=defaults[env])
-    if env_variable == "":
-        return defaults[env]
-    return env_variable
+def get_config_with_default(key: str, default_value: str) -> int:
+    value = config(key, default=default_value)
+    return int(default_value) if value.strip() == "" else int(value)
```
Comment (bbelderbos) [inline tollway/utils.py:70]: seems you can do one int() cast here

**Review by bbelderbos — COMMENTED** [review]
Good stuff, some feedback.

File: tollway/utils.py (hunk 1 above)
Comment (BenGriffith) [reply tollway/utils.py:12]: yeah, this is better

File: tollway/constants.py (hunk 3 above)
Comment (BenGriffith) [reply tollway/constants.py:11]: in the `events` module DUPLICATE_RATE and LATE_EVENT_RATE[...] are being compared to len which is why I'm casting to `int` here. Suppose I could leave this as string and then cast as `int` in the `events` module. I would also need to cast ALL_EVENTS_COUNT in `__main__` when comparison occurs. 

Screenshot 2024-02-14 at 4 38 45 PM

File: tollway/utils.py (hunk 4 above)
Comment (BenGriffith) [reply tollway/utils.py:35]: I like this idea

File: tollway/utils.py (hunk 5 above)
Comment (BenGriffith) [reply tollway/utils.py:70]: yeah, I was testing a few different scenarios:

1. No .env file
2. .env file with few of variables defined
3. .env file with all variables defined

I kept hitting value int() doesn't have .strip() attribute. I can't remember exactly.

File: tollway/utils.py [hunk 6]
```diff
@@ -1,43 +1,70 @@
...
     return json.dumps(message).encode("utf-8")
 
 
+def future_callback(logger: logging.Logger) -> Callable[[Future], None]:
+    def callback(future: Future) -> None:
+        try:
+            message_id = future.result()
+            logger.info(f"Message published with ID: {message_id}")
+        except Exception as e:
+            logger.error(f"An exception occurred: {e}")
```
[the remark this answers is not in the document]
Comment (BenGriffith) [reply tollway/utils.py:57]: very cool

File: tollway/utils.py [hunk 7]
```diff
@@ -1,43 +1,70 @@
...
 def encode_message(message: dict) -> bytes:
     return json.dumps(message).encode("utf-8")
 
 
+def future_callback(logger: logging.Logger) -> Callable[[Future], None]:
+    def callback(future: Future) -> None:
+        try:
+            message_id = future.result()
+            logger.info(f"Message published with ID: {message_id}")
+        except Exception as e:
```
[the remark this answers is not in the document]
Comment (BenGriffith) [reply tollway/utils.py:56]: there might be. I'll look into it more.

File: tollway/utils.py (hunk 5 above)
Comment (BenGriffith) [reply tollway/utils.py:70]: so I was hitting value int() type doesn't have .strip() but I think that was for something else I was trying out. I think I originally had `default_value` parameter receiving an argument of type `int`. 

Anyways, if I don't place an `int` in the else block then what is returned is a string and that won't work against the len() comparison in the `events` module. Not sure if you can see, but here's a screenshot showing DUPLICATE_RATE defined in .env as 10, remove `int` from else block and "10" is returned. 

Screenshot 2024-02-14 at 5 14 45 PM

File: tollway/constants.py (hunk 3 above)
Comment (BenGriffith) [reply tollway/constants.py:11]: yes, that's it (please see comments below related to this one). When there is no .env file, I started by using ints like `50` for default_value, however, I then was receiving an error for the value.strip() as seen in screenshot here so I changed from `int` to `str`.

Screenshot 2024-02-14 at 5 54 30 PM

Comment (BenGriffith) [reply tollway/constants.py:11]: this snippet could work as an alternative
Screenshot 2024-02-14 at 6 07 55 PM

Comment (BenGriffith) [reply tollway/constants.py:11]: scratch that suggestion. Think I'm going to go with this one - so pass an int to the default_value of the wrapper and then in the wrapper remove .strip() and do the int(value) cast. Shown in screenshot.

Screenshot 2024-02-14 at 7 32 28 PM

File: tollway/utils.py (hunk 5 above)
Comment (bbelderbos) [reply tollway/utils.py:70]: hm interesting, I remember now that config has a cast kwarg, e.g. `DEBUG = config('DEBUG', default=False, cast=bool)`, would an `cast=int` help here?

File: tollway/constants.py (hunk 3 above)
Comment (bbelderbos) [reply tollway/constants.py:11]: Maybe we can revisit this when we meet ...

**Pull request merged**